ruby/lib/google/protobuf/descriptor_dsl.rb

```diff
@@ -451,6 +451,10 @@ def value(name, number)
         end
       end
 
+      def self.decode_options(klass, serialized_options)
+        options = klass.decode(serialized_options)
+        options.freeze
```

Comment:
We want this freeze to be recursive. Otherwise a user could write something like:

```ruby
# This works, because only my_descriptor.options itself is frozen.
my_descriptor.options.some_message_options.field = 5
```

We'll want to write a loop that visits all fields and freezes them all. I'm not sure how easy this will be to do in Ruby. We may want to do it in C.

Reply:
Thank you for the review.  I've addressed the first 2 concerns (I _think_).

I have a question about the recursive freeze.  The [Message class already has a `freeze` method](https://github.com/protocolbuffers/protobuf/blob/main/ruby/ext/google/protobuf_c/message.c#L892-L906) that I am calling here.  Should that method be updated to be recursive or should I create a separate `deep_freeze` method for this use case?

Additionally, for the recursive freeze, should all fields also be pinned to the same arena?

Reply:
`Message#freeze` follows the general Ruby semantics of only doing a shallow freeze. I don't think we should change that.

I think we could add `deep_freeze` as an internal-only API, but I don't think we should make it public. We should keep new public APIs to a minimum, since each one will need to be reviewed and scrutinized.